DatadogCore/Sources/Core/Storage/FilesOrchestrator.swift

```diff
@@ -118,6 +123,26 @@ internal class FilesOrchestrator: FilesOrchestratorType {
         return newFile
     }
 
+    /// Generates a unique file name based on the current time.
+    /// The function ensures the generated file name does not already exist in the directory by checking for existing file names
+    /// and retrying with a new timestamp if a conflict is found.
+    ///
+    /// In practice, name conflicts are extremely unlikely due to the monotonic nature of `dateProvider.now`.
+    /// Conflicts could only occur in very specific scenarios, such as during a tracking consent change when files are moved
+    /// from an unauthorized (.pending) folder to an authorized (.granted) folder, with events being written immediately before
+    /// and after the consent change. These conflicts were observed in tests, causing flakiness. In real-device scenarios,
+    /// conflicts can occur if tracking consent is changed and two events are written within the precision window defined
+    /// by `Constants.fileNamePrecission` (1 millisecond).
+    private func nextFileName() -> String {
+        var newFileName = fileNameFrom(fileCreationDate: dateProvider.now)
+        while directory.hasFile(named: newFileName) {
+            // Wait for the precision duration to avoid generating the same file name
+            Thread.sleep(forTimeInterval: Constants.fileNamePrecision)
```

Comment:
I'm also not a fan of adding a sleep here. Since the date is used to sort batches at upload, can't we just add a 1ms to the `dateProvider.now` instead?

Reply:
Makes sense, I've updated the code to add `1ms` in case of a conflict 👍 ✅.

@mariedm, appending a random ID could have more indirect consequences here. On upload trigger, we read the recent batches and sort them by date. Introducing an ID would make the ordering ambiguous if two timestamps are the same, which would require additional code changes to handle these edge cases. For that reason, adjusting the timestamp seems like a less impactful solution.

Reply:
In that case I agree, using a random ID doesn't make sense.